### ISSUE TITLE
Install binaries sometimes is not complete

### ISSUE THREAD

**Issue by ilg-ul** [xpack/xpm-js#154]
In certain conditions `xpm install` does not install all dependencies:

```
2022-10-26T20:24:37.7584603Z ##[group]Run xpm install --config linux-arm64
2022-10-26T20:24:37.7585896Z [36;1mxpm install --config linux-arm64[0m
2022-10-26T20:24:37.7586848Z [36;1mxpm run build --config linux-arm64[0m
2022-10-26T20:24:37.7588170Z shell: bash --noprofile --norc -e -o pipefail {0}
2022-10-26T20:24:37.7589107Z ##[endgroup]
2022-10-26T20:24:39.2346756Z @xpack-dev-tools/cmake --config linux-arm64...
2022-10-26T20:24:39.5044552Z + @xpack-dev-tools/patchelf@0.15.0-1.1
2022-10-26T20:24:39.5077512Z + @xpack-dev-tools/pkg-config@0.29.2-1.1
2022-10-26T20:24:39.5145988Z + @xpack-dev-tools/gcc@12.2.0-1.1
2022-10-26T20:24:39.5192874Z + @xpack-dev-tools/ninja-build@1.11.1-1.1
2022-10-26T20:24:39.5220748Z + @xpack-dev-tools/cmake@3.22.6-1.1
2022-10-26T20:24:40.4267659Z @xpack-dev-tools/pkg-config@0.29.2-1.1 => '/github/home/.local/xPacks/@xpack-dev-tools/pkg-config/0.29.2-1.1'
2022-10-26T20:24:40.4349583Z Downloading https://github.com/xpack-dev-tools/pkg-config-xpack/releases/download/v0.29.2-1/xpack-pkg-config-0.29.2-1-linux-arm64.tar.gz...
2022-10-26T20:24:40.5568425Z @xpack-dev-tools/patchelf@0.15.0-1.1 => '/github/home/.local/xPacks/@xpack-dev-tools/patchelf/0.15.0-1.1'
2022-10-26T20:24:40.5659617Z Downloading https://github.com/xpack-dev-tools/patchelf-xpack/releases/download/v0.15.0-1/xpack-patchelf-0.15.0-1-linux-arm64.tar.gz...
2022-10-26T20:24:40.5883374Z @xpack-dev-tools/ninja-build@1.11.1-1.1 => '/github/home/.local/xPacks/@xpack-dev-tools/ninja-build/1.11.1-1.1'
2022-10-26T20:24:40.6143343Z Downloading https://github.com/xpack-dev-tools/ninja-build-xpack/releases/download/v1.11.1-1/xpack-ninja-build-1.11.1-1-linux-arm64.tar.gz...
2022-10-26T20:24:40.6328556Z @xpack-dev-tools/cmake@3.22.6-1.1 => '/github/home/.local/xPacks/@xpack-dev-tools/cmake/3.22.6-1.1'
2022-10-26T20:24:40.6347018Z @xpack-dev-tools/gcc@12.2.0-1.1 => '/github/home/.local/xPacks/@xpack-dev-tools/gcc/12.2.0-1.1'
2022-10-26T20:24:40.6378402Z Downloading https://github.com/xpack-dev-tools/cmake-xpack/releases/download/v3.22.6-1/xpack-cmake-3.22.6-1-linux-arm64.tar.gz...
2022-10-26T20:24:40.6502966Z Downloading https://github.com/xpack-dev-tools/gcc-xpack/releases/download/v12.2.0-1/xpack-gcc-12.2.0-1-linux-arm64.tar.gz...
2022-10-26T20:24:43.7785585Z Extracting 'xpack-patchelf-0.15.0-1-linux-arm64.tar.gz'...
2022-10-26T20:24:44.0038884Z Extracting 'xpack-ninja-build-1.11.1-1-linux-arm64.tar.gz'...
2022-10-26T20:24:44.3312336Z 34 files => '/github/home/.local/xPacks/@xpack-dev-tools/patchelf/0.15.0-1.1/.content'
2022-10-26T20:24:44.9527473Z 'xpacks/xpack-dev-tools-patchelf' -> '/github/home/.local/xPacks/@xpack-dev-tools/patchelf/0.15.0-1.1'
2022-10-26T20:24:44.9742854Z 'xpacks/.bin/patchelf' -> '../xpack-dev-tools-patchelf/.content/bin/patchelf'
2022-10-26T20:24:45.0735532Z 130 files => '/github/home/.local/xPacks/@xpack-dev-tools/ninja-build/1.11.1-1.1/.content'
2022-10-26T20:24:45.1061764Z 'xpacks/xpack-dev-tools-ninja-build' -> '/github/home/.local/xPacks/@xpack-dev-tools/ninja-build/1.11.1-1.1'
2022-10-26T20:24:45.1097429Z 'xpacks/.bin/ninja' -> '../xpack-dev-tools-ninja-build/.content/bin/ninja'
2022-10-26T20:24:46.0750190Z Extracting 'xpack-pkg-config-0.29.2-1-linux-arm64.tar.gz'...
2022-10-26T20:24:46.8730067Z 46 files => '/github/home/.local/xPacks/@xpack-dev-tools/pkg-config/0.29.2-1.1/.content'
2022-10-26T20:24:46.9128592Z 'xpacks/xpack-dev-tools-pkg-config' -> '/github/home/.local/xPacks/@xpack-dev-tools/pkg-config/0.29.2-1.1'
2022-10-26T20:24:46.9233978Z 'xpacks/.bin/pkg-config' -> '../xpack-dev-tools-pkg-config/.content/bin/pkg-config'
2022-10-26T20:24:48.7464415Z > bash  scripts/build.sh --target linux-arm64 --build-folder build/linux-arm64
```

The correct log is:

```
2022-10-27T07:50:27.2333419Z ##[group]Run xpm install --config linux-arm64
2022-10-27T07:50:27.2335458Z [36;1mxpm install --config linux-arm64[0m
2022-10-27T07:50:27.2337039Z [36;1mxpm run build --config linux-arm64[0m
2022-10-27T07:50:27.2339133Z shell: bash --noprofile --norc -e -o pipefail {0}
2022-10-27T07:50:27.2340545Z ##[endgroup]
2022-10-27T07:50:28.6742003Z @xpack-dev-tools/cmake --config linux-arm64...
2022-10-27T07:50:29.9755317Z + @xpack-dev-tools/patchelf@0.15.0-1.1
2022-10-27T07:50:29.9819481Z + @xpack-dev-tools/ninja-build@1.11.1-1.1
2022-10-27T07:50:29.9855409Z + @xpack-dev-tools/cmake@3.22.6-1.1
2022-10-27T07:50:30.2994417Z + @xpack-dev-tools/gcc@12.2.0-1.1
2022-10-27T07:50:30.5387368Z + @xpack-dev-tools/pkg-config@0.29.2-1.1
2022-10-27T07:50:30.9760107Z @xpack-dev-tools/cmake@3.22.6-1.1 => '/github/home/.local/xPacks/@xpack-dev-tools/cmake/3.22.6-1.1'
2022-10-27T07:50:30.9766886Z @xpack-dev-tools/gcc@12.2.0-1.1 => '/github/home/.local/xPacks/@xpack-dev-tools/gcc/12.2.0-1.1'
2022-10-27T07:50:30.9828027Z Downloading https://github.com/xpack-dev-tools/cmake-xpack/releases/download/v3.22.6-1/xpack-cmake-3.22.6-1-linux-arm64.tar.gz...
2022-10-27T07:50:30.9907331Z Downloading https://github.com/xpack-dev-tools/gcc-xpack/releases/download/v12.2.0-1/xpack-gcc-12.2.0-1-linux-arm64.tar.gz...
2022-10-27T07:50:31.0252902Z @xpack-dev-tools/pkg-config@0.29.2-1.1 => '/github/home/.local/xPacks/@xpack-dev-tools/pkg-config/0.29.2-1.1'
2022-10-27T07:50:31.0284378Z Downloading https://github.com/xpack-dev-tools/pkg-config-xpack/releases/download/v0.29.2-1/xpack-pkg-config-0.29.2-1-linux-arm64.tar.gz...
2022-10-27T07:50:31.6003669Z @xpack-dev-tools/ninja-build@1.11.1-1.1 => '/github/home/.local/xPacks/@xpack-dev-tools/ninja-build/1.11.1-1.1'
2022-10-27T07:50:31.6089466Z Downloading https://github.com/xpack-dev-tools/ninja-build-xpack/releases/download/v1.11.1-1/xpack-ninja-build-1.11.1-1-linux-arm64.tar.gz...
2022-10-27T07:50:31.6138199Z @xpack-dev-tools/patchelf@0.15.0-1.1 => '/github/home/.local/xPacks/@xpack-dev-tools/patchelf/0.15.0-1.1'
2022-10-27T07:50:31.6178143Z Downloading https://github.com/xpack-dev-tools/patchelf-xpack/releases/download/v0.15.0-1/xpack-patchelf-0.15.0-1-linux-arm64.tar.gz...
2022-10-27T07:50:31.9277010Z Extracting 'xpack-pkg-config-0.29.2-1-linux-arm64.tar.gz'...
2022-10-27T07:50:34.7276546Z Extracting 'xpack-ninja-build-1.11.1-1-linux-arm64.tar.gz'...
2022-10-27T07:50:34.8475324Z Extracting 'xpack-cmake-3.22.6-1-linux-arm64.tar.gz'...
2022-10-27T07:50:34.9397776Z Extracting 'xpack-patchelf-0.15.0-1-linux-arm64.tar.gz'...
2022-10-27T07:50:35.4418515Z 46 files => '/github/home/.local/xPacks/@xpack-dev-tools/pkg-config/0.29.2-1.1/.content'
2022-10-27T07:50:40.3260293Z 'xpacks/xpack-dev-tools-pkg-config' -> '/github/home/.local/xPacks/@xpack-dev-tools/pkg-config/0.29.2-1.1'
2022-10-27T07:50:40.3747838Z 'xpacks/.bin/pkg-config' -> '../xpack-dev-tools-pkg-config/.content/bin/pkg-config'
2022-10-27T07:50:40.5885873Z 34 files => '/github/home/.local/xPacks/@xpack-dev-tools/patchelf/0.15.0-1.1/.content'
2022-10-27T07:50:40.5896029Z 130 files => '/github/home/.local/xPacks/@xpack-dev-tools/ninja-build/1.11.1-1.1/.content'
2022-10-27T07:50:40.6505358Z 'xpacks/xpack-dev-tools-patchelf' -> '/github/home/.local/xPacks/@xpack-dev-tools/patchelf/0.15.0-1.1'
2022-10-27T07:50:40.6545981Z 'xpacks/.bin/patchelf' -> '../xpack-dev-tools-patchelf/.content/bin/patchelf'
2022-10-27T07:50:40.7575187Z 'xpacks/xpack-dev-tools-ninja-build' -> '/github/home/.local/xPacks/@xpack-dev-tools/ninja-build/1.11.1-1.1'
2022-10-27T07:50:40.8125894Z 'xpacks/.bin/ninja' -> '../xpack-dev-tools-ninja-build/.content/bin/ninja'
2022-10-27T07:50:43.1672861Z Extracting 'xpack-gcc-12.2.0-1-linux-arm64.tar.gz'...
2022-10-27T07:50:59.1338820Z 3206 files => '/github/home/.local/xPacks/@xpack-dev-tools/cmake/3.22.6-1.1/.content'
2022-10-27T07:51:00.6780230Z 'xpacks/xpack-dev-tools-cmake' -> '/github/home/.local/xPacks/@xpack-dev-tools/cmake/3.22.6-1.1'
2022-10-27T07:51:00.6824715Z 'xpacks/.bin/ccmake' -> '../xpack-dev-tools-cmake/.content/bin/ccmake'
2022-10-27T07:51:00.6849581Z 'xpacks/.bin/cmake' -> '../xpack-dev-tools-cmake/.content/bin/cmake'
2022-10-27T07:51:00.6899509Z 'xpacks/.bin/cpack' -> '../xpack-dev-tools-cmake/.content/bin/cpack'
2022-10-27T07:51:00.6914295Z 'xpacks/.bin/ctest' -> '../xpack-dev-tools-cmake/.content/bin/ctest'
2022-10-27T07:51:14.7592441Z 2127 files => '/github/home/.local/xPacks/@xpack-dev-tools/gcc/12.2.0-1.1/.content'
2022-10-27T07:51:15.1755535Z 'xpacks/xpack-dev-tools-gcc' -> '/github/home/.local/xPacks/@xpack-dev-tools/gcc/12.2.0-1.1'
2022-10-27T07:51:15.1799471Z 'xpacks/.bin/addr2line' -> '../xpack-dev-tools-gcc/.content/bin/addr2line'
2022-10-27T07:51:15.1822978Z 'xpacks/.bin/ar' -> '../xpack-dev-tools-gcc/.content/bin/ar'
2022-10-27T07:51:15.1856256Z 'xpacks/.bin/as' -> '../xpack-dev-tools-gcc/.content/bin/as'
2022-10-27T07:51:15.1884558Z 'xpacks/.bin/c++' -> '../xpack-dev-tools-gcc/.content/bin/c++'
2022-10-27T07:51:15.1918169Z 'xpacks/.bin/c++filt' -> '../xpack-dev-tools-gcc/.content/bin/c++filt'
2022-10-27T07:51:15.1969213Z 'xpacks/.bin/cpp' -> '../xpack-dev-tools-gcc/.content/bin/cpp'
2022-10-27T07:51:15.2001966Z 'xpacks/.bin/dwp' -> '../xpack-dev-tools-gcc/.content/bin/dwp'
2022-10-27T07:51:15.2056004Z 'xpacks/.bin/elfedit' -> '../xpack-dev-tools-gcc/.content/bin/elfedit'
2022-10-27T07:51:15.2099741Z 'xpacks/.bin/g++' -> '../xpack-dev-tools-gcc/.content/bin/g++'
2022-10-27T07:51:15.2130114Z 'xpacks/.bin/gcc' -> '../xpack-dev-tools-gcc/.content/bin/gcc'
2022-10-27T07:51:15.2192403Z 'xpacks/.bin/gcc-ar' -> '../xpack-dev-tools-gcc/.content/bin/gcc-ar'
2022-10-27T07:51:15.2222769Z 'xpacks/.bin/gcc-nm' -> '../xpack-dev-tools-gcc/.content/bin/gcc-nm'
2022-10-27T07:51:15.2256072Z 'xpacks/.bin/gcc-ranlib' -> '../xpack-dev-tools-gcc/.content/bin/gcc-ranlib'
2022-10-27T07:51:15.2283287Z 'xpacks/.bin/gcore' -> '../xpack-dev-tools-gcc/.content/bin/gcore'
2022-10-27T07:51:15.2307960Z 'xpacks/.bin/gcov' -> '../xpack-dev-tools-gcc/.content/bin/gcov'
2022-10-27T07:51:15.2334795Z 'xpacks/.bin/gcov-dump' -> '../xpack-dev-tools-gcc/.content/bin/gcov-dump'
2022-10-27T07:51:15.2357896Z 'xpacks/.bin/gcov-tool' -> '../xpack-dev-tools-gcc/.content/bin/gcov-tool'
2022-10-27T07:51:15.2381314Z 'xpacks/.bin/gdb' -> '../xpack-dev-tools-gcc/.content/bin/gdb'
2022-10-27T07:51:15.2404039Z 'xpacks/.bin/gdb-add-index' -> '../xpack-dev-tools-gcc/.content/bin/gdb-add-index'
2022-10-27T07:51:15.2426773Z 'xpacks/.bin/gprof' -> '../xpack-dev-tools-gcc/.content/bin/gprof'
2022-10-27T07:51:15.2448680Z 'xpacks/.bin/ld' -> '../xpack-dev-tools-gcc/.content/bin/ld'
2022-10-27T07:51:15.2477414Z 'xpacks/.bin/ld.bfd' -> '../xpack-dev-tools-gcc/.content/bin/ld.bfd'
2022-10-27T07:51:15.2499225Z 'xpacks/.bin/ld.gold' -> '../xpack-dev-tools-gcc/.content/bin/ld.gold'
2022-10-27T07:51:15.2521715Z 'xpacks/.bin/lto-dump' -> '../xpack-dev-tools-gcc/.content/bin/lto-dump'
2022-10-27T07:51:15.2544969Z 'xpacks/.bin/nm' -> '../xpack-dev-tools-gcc/.content/bin/nm'
2022-10-27T07:51:15.2568036Z 'xpacks/.bin/objcopy' -> '../xpack-dev-tools-gcc/.content/bin/objcopy'
2022-10-27T07:51:15.2643860Z 'xpacks/.bin/objdump' -> '../xpack-dev-tools-gcc/.content/bin/objdump'
2022-10-27T07:51:15.2675790Z 'xpacks/.bin/ranlib' -> '../xpack-dev-tools-gcc/.content/bin/ranlib'
2022-10-27T07:51:15.2693017Z 'xpacks/.bin/readelf' -> '../xpack-dev-tools-gcc/.content/bin/readelf'
2022-10-27T07:51:15.2762149Z 'xpacks/.bin/size' -> '../xpack-dev-tools-gcc/.content/bin/size'
2022-10-27T07:51:15.2784112Z 'xpacks/.bin/strings' -> '../xpack-dev-tools-gcc/.content/bin/strings'
2022-10-27T07:51:15.2807209Z 'xpacks/.bin/strip' -> '../xpack-dev-tools-gcc/.content/bin/strip'
2022-10-27T07:51:16.2313193Z > bash  scripts/build.sh --target linux-arm64 --build-folder build/linux-arm64
```

As it can be seen, the gcc & cmake binaries are not installed.

**Comment by ilg-ul** [comment]
The problem seems to be in getting the buffer from fetch:

```
        const data = await res.buffer()
        log.trace(`buffer for ${url} ${data.length} bytes`)
        return cacache.put(cacheFolderPath, key, data, opts)
```

More trace messages added to confirm this

**Comment by ilg-ul** [comment]
It should have been `return await cacache.put(cacheFolderPath, key, data, opts)`.

**Comment by ilg-ul** [comment]
The culprit is `ECONNRESET` which is occasionally thrown by the servers.

The above code (https://github.com/xpack/xpm-js/issues/154#issuecomment-1369526498) seems not able to throw the error correctly, not to mention that it requires huge amounts of memory.

The code was updated to use the pipeline and stream solution:

```
      const cacacheWriteStream = cacache.put.stream(cacheFolderPath, key, opts)
      await pipelinePromise(response.body, cacacheWriteStream)
```

With the current code the error is always thrown, and there are no runs with incomplete dependencies.

**Comment by ilg-ul** [comment]
Fixed in 0.14.8 on 2022-02-03.
